### ISSUE TITLE
update to React 15.4.0

### ISSUE THREAD

**Issue by yogthos** [reagent-project/reagent#275]
It doesn't look like there are any API changes, but it needs some testing.

**Comment by Deraen** [comment]
New version up in Cljsjs in a bit, and yeah, this will work BUT using react-dom-server in 15.4.0 will load considerable amount of unnecessary and duplicated code to client. Details in a moment.

From Changelog:

> Another thing to watch out for is that React DOM Server is now about the same size as React DOM since it contains its own copy of the React reconciler. We don't recommend using React DOM Server on the client in most cases.

**Comment by Deraen** [comment]
15.4.0:
- react.inc.js 125KB (21KB)
- react-dom.inc.js 605KB (122KB)
- react-dom-server.inc.js 530KB (117KB)
- total 1260KB (260KB)
- without dom-server 730KB (143KB)

15.3.1:
- react.inc.js 686KB (146KB)
- react-dom.inc.js 1KB (1KB)
- react-dom-server.inc.js 1KB (1KB)
- total 688KB (148KB)

**Comment by Deraen** [comment]
It is not necessary to drop the dependency to `cljsjs/react-dom-server`, it is enough to only require `cljsjs.react.dom.server` namespace when it is absolutely required: when `render-to-string` or `render-to-static-markup` is needed, so those must be removed from core namespace and users need to start using `reagent.dom.server` when they need those. This is a breaking change, but IMO this makes sense.

**Comment by vans163** [comment]
React 15.4.1 gives error Uncaught Error: js/ReactDOM is missing. This now became js/React.DOM.

This happens when you use an external version of react passing :exclusions [cljsjs/react]

**Comment by Deraen** [comment]
@vans163 Not true. ReactDOM is still ReactDOM. React.DOM is wrapper for createElement:

> Convenience wrappers around React.createElement() for DOM components are provided by React.DOM. For example, React.DOM.a(...) is a convenience wrapper for React.createElement('a', ...).

You are probably missing the separate react-dom file.

**Comment by vans163** [comment]
```html
  <script src="https://unpkg.com/react@15.4.1/dist/react-with-addons.js"></script>
  <script src="https://unpkg.com/react-dom@15.4.1/dist/react-dom.js"></script>
```

If I change the version to @15.3.2 ReactDOM is there. otherwise its React.DOM.

Am I doing something wrong here?

**Comment by Deraen** [comment]
Works for me.

Did you also remember to exclude `cljsjs.react-dom`?

What are you using to provide `cljsjs.react` and `cljsjs.react.dom` namespaces? Empty cljs namespaces should work.

**Comment by vans163** [comment]
[reagent "0.6.0" :exclusions [cljsjs/react]] inside the project.clj dependencies.

So I just need to make this [reagent "0.6.0" :exclusions [cljsjs/react cljsjs/react-dom]] ? Because this doesnt work and gives tons of clojurescript errors.

**Comment by Deraen** [comment]
Check `lein deps` or `boot show -d` to check the dependency tree. It is possible that you have a other library pulling in Cljsjs/react dependency. Cljs errors are expected if you exclude cljsjs.react but don't provide the mentioned namespaces yourself.

If you have other libraries pulling in cljsjs/react & dom transitively, you can use global exclusions.

**Comment by vans163** [comment]
leins deps returns empty.

I have [reagent "0.6.0" :exclusions [cljsjs/react cljsjs/react-dom]] and two files with namespaces (ns cljsjs.react) and (ns cljsjs.react-dom)

I deleted the entire output dir and recompiled everything now I get:
```
ANALYSIS ERROR: No such namespace: cljsjs.react.dom, could not locate cljsjs/react/dom.cljs, cljsjs/react/dom.cljc, or Closure namespace "cljsjs.react.dom" in file /out/reagent/dom.cljs on file null, line null, column null
```

**Comment by Deraen** [comment]
@vans163 Yes, as I said that is expected. You removed cljsjs dependencies so now the namespace required by Reagent is not available. You need to provide that namespace yourself. You can for example create empty cljs namespace with that name.

**Comment by vans163** [comment]
@Deraen Got it working. I think.. so far.

The homepage says
```
[reagent "0.6.0" :exclusions [cljsjs/react]]
[cljsjs/react-with-addons "15.2.1-0"]

If you want to use your own build of React (or React from a CDN), you have to use :exclusions variant of the dependency, and also provide a file named "cljsjs/react.cljs", containing just (ns cljsjs.react), in your project.
```

So what do I need to put in the project dependencies? 
```
[reagent "0.6.0" :exclusions [cljsjs/react cljsjs/react-dom]]
```

And thats it? I am using react-with-addons.  If I change react to react-with-addons I cannot create a namespace called react.with.addons.

**Comment by Deraen** [comment]
@vans163 Even if you use react-with-addons, you need to provide `cljsjs.react` namespace. This is because both cljsjs/react and cljsjs/react-with-addons provide the same namespace.

If you use cljsjs/react-with-addons, you don't need to exclude cljsjs/react-dom, because you would still use react-dom.

The readme is a bit outdated on this part, I think following should be up-to-date guide.

> If you want to use your own build of React (or React from a CDN), you have to use :exclusions to remove `cljsjs/react`, `cljsjs/react-dom` and `cljsjs/react-dom-server` and also provide empty Cljs namespaces `cljsjs.react` and `cljsjs.react.dom` for Reagent.

**Comment by vans163** [comment]
@Deraen Yea this works and makes sense now.

**Comment by Deraen** [comment]
Fixed in 0.6.1.
